pkg/common/health/health.go

```diff
@@ -0,0 +1,95 @@
+package health
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/InVisionApp/go-health"
+	"github.com/InVisionApp/go-health/handlers"
+	"github.com/sirupsen/logrus"
+)
+
+// health.Checker is responsible for running health checks and serving the healthcheck HTTP paths
+type Checker struct {
+	config Config
+
+	server *http.Server
+
+	hc *health.Health
+
+	log logrus.FieldLogger
+}
+
+func hello(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(http.StatusOK)
+}
+
+func NewChecker(config Config, log logrus.FieldLogger) *Checker {
+	hc := health.New()
+
+	var server http.Server
+	// Start HTTP server if address is configured
+	if config.Address != "" {
+		handler := http.NewServeMux()
+
+		handler.HandleFunc(config.getHealthPath(), handlers.NewJSONHandlerFunc(hc, nil))
+		handler.HandleFunc(config.getLivePath(), hello)
+
+		server = http.Server{
+			Addr:    config.Address,
+			Handler: handler,
+		}
+	}
+
+	hc.StatusListener = &statusListener{}
+
+	return &Checker{config: config, server: &server, hc: hc, log: log}
+}
+
+func (c *Checker) AddCheck(name string, checker health.ICheckable) error {
+	return c.hc.AddCheck(&health.Config{
+		Name:     name,
+		Checker:  checker,
+		Interval: c.config.getInterval(),
+		Fatal:    true,
+	})
+}
+
+func (c *Checker) ListenAndServe(ctx context.Context) error {
+	if err := c.hc.Start(); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	if c.server != nil {
+		go func() {
+			defer wg.Done()
+			c.log.WithFields(logrus.Fields{
+				"address": c.config.Address, "interval": c.config.getInterval(),
+			}).Info("Serving health checks")
+			if err := c.server.ListenAndServe(); err != http.ErrServerClosed {
+				c.log.WithError(err).Warn("Error serving health checks")
+			}
+		}()
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-ctx.Done()
+		if c.server != nil {
+			c.server.Shutdown(ctx)
```

Comment:
Context will already be cancelled at this point - perhaps we should call Close instead?

Reply:
done